### PR TITLE
Fix #7559: autodoc: misdetects a sync function is async

### LINKED ISSUE #7559: Sphix indicates that a sync function is async
This one is maybe my fault, but I'm not sure what I should be doing.

**Describe the bug**
I'm generating async methods from sync methods this way:
```python
    @functools.wraps(wrapped)
    def wrapper(*args, **kwargs):
        awaitable = wrapped(*args, **kwargs)
        return sync_await(awaitable=awaitable)
```
Because I use ``functools.wraps`` and after https://github.com/sphinx-doc/sphinx/pull/7222, Sphinx sees that my sync function wraps an async function and unwraps it in ``iscoroutinefunction``. the wrapped function is async, so the function is async.

When documenting, both the original and wrapped functions are seen as async.

**To Reproduce**
```
$ git clone https://github.com/peopledoc/procrastinate
$ cd procrastinate
$ pip install tox
$ tox -e docs
$ # open docs/_build/html/reference.html#procrastinate.tasks.Task.defer, it's noted as async
```

**Expected behavior**
Not really sure. For a staticmethod, for example, it makes sense to consider it async even if it's the wrapped method that is async.

**Your project**
https://github.com/peopledoc/procrastinate
https://github.com/peopledoc/procrastinate/blob/9dd6c296ddf59bb524e20cd6bd09703ff00a8f66/procrastinate/utils.py

**Screenshots**
Capture d’écran 2020-04-26 à 14 58 22

**Environment info**
- OS: All
- Python version: 3.7.7
- Sphinx version: Sphinx==3.0.2
- Sphinx extensions:
```
extensions = [
    "sphinx.ext.napoleon",
    "sphinx.ext.autodoc",
    "sphinx_autodoc_typehints",
    "sphinxcontrib.programoutput",
    "changelog",
]
```

**Additional context**
I'm not sure is this is because I shouldn't be using wraps if I expect the wrapper and the wrapped function to be seen as distinct, or if I'm doing something unexpected. I know I'm probably doing something weird in the first place with automatic synchronous functions generation, so if you think this is out of scope, I understand.

If there's a workaround, though, I'm interested.

### PR DESCRIPTION
**Pull request by tk0miya**
### Feature or Bugfix
- Bugfix

### Purpose
- refs: #7559

### CODE REVIEW

**Comment by eric-wieser** [comment]
What's the expected behavior of the following?

```python
def wrap_it(f):
   @wraps(f)
   def wrapper(*args, **kwargs):
       return f(*args, **kwargs)
   return wrapper

# async or not?
async_wrapped = wrap_it(async_func)
```

**Comment by tk0miya** [comment]
For now, `inspect.iscoroutinefunction()` considers it is not a coroutine-function. So I followed the behavior. I know the original meaning of coroutine-function is a function that returns coroutine object. But it is a bit difficult to inspect it without the "async" keyword.

```
def my_coroutine_func():
    """This is coroutine function too"""
    async def get_coroutine_obj():
        pass

    return get_coroutine_obj()
```

**Comment by eric-wieser** [comment]
Seems fair, thanks for explaining.

File: sphinx/util/inspect.py [hunk 1]
```diff
@@ -125,13 +125,15 @@ def unwrap(obj: Any) -> Any:
...
+def unwrap_all(obj: Any, *, stop: Callable = None) -> Any:
     """
     Get an original object from wrapped object (unwrapping partials, wrapped
     functions, and other decorators).
     """
     while True:
-        if ispartial(obj):
+        if stop and stop(obj):
+            return obj
+        elif ispartial(obj):
```
Comment (eric-wieser) [inline sphinx/util/inspect.py:136]: Can you add a test of a partial async function?

Comment (tk0miya) [reply sphinx/util/inspect.py:136]: We already have it 
https://github.com/sphinx-doc/sphinx/blob/017af4ce2ef1fbc4387100b850e202b17f4c7450/tests/test_util_inspect.py#L473

**Comment by tk0miya** [comment]
Merged. @eric-wieser Thank you for reviewing!

**Comment by ewjoachim** [comment]
(Ticket OP here, thanks @eric-wieser too !)

**Pull request merged**